Passing user roles from query originator to other nodes

Enmk opened the pull request:
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Push user roles from query originator to other nodes in cluster. Helpful when only originator has access to the external authenticator (like LDAP).
...

closes: #34130

Enmk committed:
Passing user roles from query originator to other nodes

This is to address case when originator have access to remote
auth provider (like LDAP server), but other nodes in cluster don't.

alexey-milovidov commented:
Missing tests.

vitlibar commented:
Is this PR still in work?

vitlibar commented on src/Access/AccessControl.h:
```diff
@@ -169,6 +169,7 @@ class AccessControl : public MultipleAccessStorage
     std::shared_ptr<const ContextAccess> getContextAccess(
         const UUID & user_id,
         const std::vector<UUID> & current_roles,
+        const std::vector<UUID> & additional_roles,
```
Why can't we combine those `additional_roles` with `current_roles` and use one parameter instead of two ones?

Enmk replied:
Because there is also `use_default_roles`, that controls which roles are applied to the `ContextAccess`.

I'm not entirely sure why `use_default_roles` should be `false` or `true`, hence wasn't sure if the same logic should be applied to `external_roles`.

vitlibar commented on src/Access/AccessControl.h:
```diff
@@ -169,6 +169,7 @@ class AccessControl : public MultipleAccessStorage
     std::shared_ptr<const ContextAccess> getContextAccess(
         const UUID & user_id,
         const std::vector<UUID> & current_roles,
+        const std::vector<UUID> & additional_roles,
```
What are those `additional_roles` used for?

vitlibar commented on src/Access/AccessControl.cpp:
```diff
@@ -447,8 +447,13 @@ UUID AccessControl::authenticate(const Credentials & credentials, const Poco::Ne
     {
-        return MultipleAccessStorage::authenticate(credentials, address, *external_authenticators, allow_no_password,
+        LOG_DEBUG(getLogger(), "AccessControl::authenticate {} user {}",
+                typeid(credentials).name(), credentials.getUserName());
```
Isn't it enough logging for authentication already?

Enmk replied:
Those are leftovers from debugging session and should be removed

vitlibar commented on src/Access/ContextAccess.cpp:
```diff
@@ -708,7 +708,7 @@ bool ContextAccess::checkAdminOptionImplHelper(const Container & role_ids, const
             auto role_name = get_name_function(role_id, i);
             if (!role_name)
-                role_name = "ID {" + toString(role_id) + "}";
+                role_name = "ID {" + DB::toString(role_id) + "}";
```
What was wrong with just `toString`?

vitlibar commented on src/Access/IAccessStorage.cpp:
```diff
@@ -4,6 +4,7 @@
 #include <Access/User.h>
 #include <Access/AccessBackup.h>
 #include <Backups/BackupEntriesCollector.h>
+#include "Common/logger_useful.h"
```
why?

Enmk replied:
Leftovers from debugging, will be removed.

vitlibar commented on src/Interpreters/Session.h:
```diff
@@ -90,6 +93,7 @@ class Session
 
     mutable UserPtr user;
     std::optional<UUID> user_id;
+    std::vector<UUID> extra_granted_roles;
```
If I understood correctly, `additional roles` == `extra_roles` == `extra_granted_roles`? Is it correct?

Enmk replied:
Agree, renaming all that to just `external_roles`

vitlibar commented on src/QueryPipeline/RemoteQueryExecutor.cpp:
```diff
@@ -249,7 +254,19 @@ void RemoteQueryExecutor::sendQuery(ClientInfo::QueryKind query_kind)
     }
 
-    connections->sendQuery(timeouts, query, query_id, stage, modified_client_info, true);
+    auto user = context->getAccessControl().read<User>(modified_client_info.initial_user, true);
+    boost::container::flat_set<String> granted_roles;
+    if (user)
+    {
+        const auto & access_control = context->getAccessControl();
+        for (const auto & e : user->granted_roles.getElements())
+        {
+            auto names = access_control.readNames(e.ids);
+            granted_roles.insert(names.begin(), names.end());
+        }
+    }
```
This part looks strange? What is it for?

Enmk replied:
Roles UUIDs might be different on cluster nodes, while names 100% match, so it makes sense to send roles as list of names rather than list of UUIDs.

vitlibar commented on src/Interpreters/Session.cpp:
```diff
@@ -307,14 +309,22 @@ void Session::authenticate(const Credentials & credentials_, const Poco::Net::So
+        LOG_DEBUG(log, "Authenticated with global context as user {}",
+                toString(*user_id));
+
+        if (!extra_granted_roles_.empty() && global_context->getSettingsRef().allow_extenral_roles_in_interserver_queries)
```
Probably it's better to check this setting before sending those external roles (and don't send them if it's false).

Enmk replied:
yep, please see https://github.com/ClickHouse/ClickHouse/pull/42537/files#diff-a05838cce3aacc32a5b2fc017af198fb3086277d8912159aa7a3dcac7aeb7478R260

Enmk committed:
Fixes according to PR review

Sending roles only if `allow_extenral_roles_in_interserver_queries` setting is ON
renamed various naming variants of 'additional', 'extra', 'externally granted' to 'external_roles'
removed various debugging leftovers
minor cleanup

alexey-milovidov commented:
Missing tests.

Enmk committed:
Merge branch 'master' into ldap_remote_roles

Enmk committed:
Fixed typo

Enmk committed:
Removed extra include

Enmk committed:
Minor cleanup

Enmk commented:
> Missing tests.

Could you please mark this as "ready for tests", so we can execute produced binaries against LDAP suite from here: https://github.com/Altinity/clickhouse-regression/tree/main/ldap ?

alexey-milovidov commented:
The tests should run in CI.

zvonand committed:
upload not working

zvonand committed:
fixed, shortened

Enmk committed:
removed ldap image from list

Since it is not meant to be rebuilt by the CI/CD

Enmk committed:
Merge remote-tracking branch 'upstream/master' into ldap_remote_roles

robot-clickhouse commented:
<!-- automatic status comment for PR #42537 from Enmk/ClickHouse:ldap_remote_roles -->
This is an automated comment for commit 558c806bfd66449f04e9ea686b898edcec22aae3 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/42537/558c806bfd66449f04e9ea686b898edcec22aae3/ci_running.html)
The overall status of the commit is 🟢 success

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>Style Check</td><td>Runs a set of checks to keep the code style clean. If some of tests failed, see the related log from the report</td><td>🟢 success</td></tr>
</table>

Enmk committed:
Merge pull request #5 from zvonand/zvonand-ldap-tests

Tests for PR 42537

Enmk committed:
Update dockerd-entrypoint.sh

Enmk committed:
Merge branch 'master' into ldap_remote_roles

Enmk commented:
@vitlibar please take a look

Enmk committed:
Merge branch 'master' into ldap_remote_roles

Enmk committed:
Merge branch 'master' into ldap_remote_roles

Enmk commented:
@vitlibar please review

Enmk committed:
Merge branch 'master' into ldap_remote_roles

vitlibar commented:
Please fix the conflicts

vitlibar commented on src/Access/LDAPAccessStorage.cpp:
```diff
@@ -320,6 +319,10 @@ std::set<String> LDAPAccessStorage::mapExternalRolesNoLock(const LDAPClient::Sea
+        return role_names;
+
     if (external_roles.size() != role_search_params.size())
         throw Exception(ErrorCodes::BAD_ARGUMENTS, "Unable to map external roles");
```
What if it gets a list of external roles from the initiator, but the configuration file is not the same here so `role_search_params` is not the same here as it was on the initiator and we fail here? I mean maybe this check and also the check for role name's prefix is useful here in general but not for external roles from the initiator?

vitlibar commented on src/Server/TCPHandler.cpp:
```diff
@@ -1496,6 +1500,13 @@ void TCPHandler::receiveQuery()
 
+    std::string received_extra_roles;
+    // TODO: check if having `is_interserver_mode` doesn't break interoperability with the CH-client.
+    if (client_tcp_protocol_version >= DBMS_MIN_PROTOCOL_VERSION_WITH_INTERSERVER_EXTERNALLY_GRANTED_ROLES)
```
`DBMS_MIN_PROTOCOL_VERSION_WITH_INTERSERVER_EXTERNAL_ROLES` is a bit shorter

vitlibar commented on src/Access/ContextAccess.cpp:
```diff
@@ -299,6 +299,7 @@ void ContextAccess::setUser(const UserPtr & user_) const
         current_roles = user->granted_roles.findGranted(params.current_roles);
         current_roles_with_admin_option = user->granted_roles.findGrantedWithAdminOption(params.current_roles);
     }
+    current_roles.insert(current_roles.end(), params.external_roles.begin(), params.external_roles.end());
```
It doesn't feel right to add roles to `current_roles` without adding them to `user->default_roles`. I think we can just put external roles instead of current roles, and modify `User` in `LDAPAccessStorage` to contain the default roles we want.

vitlibar commented on src/Access/AccessControl.h:
```diff
@@ -184,6 +184,7 @@ class AccessControl : public MultipleAccessStorage
     std::shared_ptr<const ContextAccess> getContextAccess(
         const UUID & user_id,
         const std::vector<UUID> & current_roles,
+        const std::vector<UUID> & external_roles,
```
And that would also allow another server to understand changes in current roles. For example,

```
SET CURRENT ROLE role1, role2;
SELECT * FROM remote(another server);
```

so another server would understand that current roles are `role1` and `role2` even if the initial user has more granted roles.

vitlibar commented on src/QueryPipeline/RemoteQueryExecutor.cpp:
```diff
@@ -263,7 +266,25 @@ void RemoteQueryExecutor::sendQueryUnlocked(ClientInfo::QueryKind query_kind, As
+            }
+        }
+        local_granted_roles.insert(local_granted_roles.end(), granted_roles.begin(), granted_roles.end());
+    }
```
And also it's much easier here to extract current roles instead of collecting granted roles:

```
auto current_roles_names = getRolesInfo()->getCurrentRolesNames();
connections->sendQuery(timeouts, query, query_id, stage, modified_client_info, true, current_roles_names);
```

vitlibar commented:
https://github.com/ClickHouse/ClickHouse/pull/52660

alexey-milovidov commented:
> Please fix the conflicts

Please reopen after fixing the conflicts.
